initial gnss derived tec decoder

BenjaminRuston opened the pull request:
build-group=https://github.com/JCSDA-internal/ioda/pull/1256


## Description

This decodes a COSMIC-2 TEC file and puts it into IODA format.

This is related to the IODA PR:
https://github.com/JCSDA-internal/ioda/pull/1256

## Issue(s) addressed

Resolves #1504 

## Dependencies

List the other PRs that this PR is dependent on:
- [x] https://github.com/JCSDA-internal/ioda/pull/1256

## Impact

Initial capability for TEC decoding from netCDF to IODA

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have run the unit tests before creating the PR

BenjaminRuston committed:
initial commit of initial decoder non functional at this time

BenjaminRuston commented:
@davidkuhl this is just a start, but see if I can get a functional decoder soon, could have short cut it but trying to be generally thorough in the creation of the initial prototype

would be good to get other TEC files from potentially other constellations beyond COSMIC-2 and make sure this is the latest format of the files

BenjaminRuston commented on src/gnssro/gnss_tec_bufr2ioda.py:
```diff
@@ -0,0 +1,239 @@
+    meta_data_types = {
+        "dateTime": "long",
+        "elevationAngleGNSSlink": "float",
+        "GNSSxECFPosition": "float",
```
@davidkuhl I'm coming up with these new IODA entry names, but we'll need to but a bit more formal and likely should ask NOAA SWPC but trying to make this a bit more descriptive and also generic so `x_GPS` now `GNSSxECFPosition` and `x_LEO` now `SATxECFPosition`

davidkuhl replied:
Let me see if Victoriya has any ideas on more descriptive names for these.

BenjaminRuston replied:
@davidkuhl this is not ready yet... ;0)

will let you know when I generate a file, we will also bring in other Victoriya for certain to discuss convention

BenjaminRuston commented on src/gnssro/gnss_tec_bufr2ioda.py:
```diff
@@ -0,0 +1,239 @@
+            aelev = np.concatenate((aelev, elev), axis=None)
+            aUT_RO = np.concatenate((aUT_RO, UT), axis=None)
+            aRO_ind = np.concatenate((aRO_ind, RO_ind), axis=None)
+            aunique_number = np.concatenate((aunique_number, unique_number), axis=None)
```
need to change this to `sequenceNumber`

BenjaminRuston commented:
adding letter to number follow WMO BUFR:
![image](https://github.com/JCSDA-internal/ioda-converters/assets/93737224/53b21adf-e094-4813-bf5a-64a900facfcf)

BenjaminRuston committed:
complete initial working prototype

BenjaminRuston committed:
fix more metaData issues still somewhat functional

BenjaminRuston committed:
update sequenceNumber metaData for multiple files

BenjaminRuston committed:
fix coding norms

BenjaminRuston committed:
fix WMO BUFR numbers for tranmsmitter and satelliteID

fcvdb reviewed (approved):
The converter was successfully tested on Narwhal and Nautilus with file `podTc2_C2E3.2020.275.08.18.0013.G10.01_0001.0001_nc`

PatNichols committed:
Merge branch 'develop' into feature/initial_tec_decoder

BenjaminRuston committed:
add computation of latitude and longitude from ECF coordinates

BenjaminRuston commented:
@davidkuhl so I've added a function to compute the semi-accurate lat/lon from the ECF coordinates.  Think this will be helpful as now the IODA file will contain the same `required` data as the other data types.  The operator is fully welcome to not use these and proceed to use the ECF coordinates.  

I've started the IODA PR too, added to this PR description think have a good proposal to go forward with

BenjaminRuston committed:
update name of angle to GNSS transmitter

BenjaminRuston commented on src/gnssro/gnss_tec_bufr2ioda.py:
```diff
@@ -0,0 +1,375 @@
+    ('elevationAngleGNSS', 'float'),
+    ('latitude', 'float'),
+    ('longitude', 'float'),
+    ('xECFPosition', 'float'),
```
```suggestion
    ('xECEFPosition', 'float'),
```

when I look up this coordinate system `ECEF` is more commonly used, should we adopt this as the standard?

BenjaminRuston commented on src/gnssro/gnss_tec_bufr2ioda.py:
```diff
@@ -0,0 +1,375 @@
+      f = 1 / 298.257
+
+    Input
+      x: coordinate X meters
```
```suggestion
      x: coordinate x meters
```

BenjaminRuston commented:
@davidkuhl and Victoriya the xyz2llh was abstracted from here:
https://gis.stackexchange.com/questions/265909/converting-from-ecef-to-geodetic-coordinates

BenjaminRuston commented:
these data will use the international reference ionosphere and an interface will develop with this version:
https://github.com/JCSDA-internal/pyiri-jedi

BenjaminRuston committed:
Update src/gnssro/gnss_tec_bufr2ioda.py

BenjaminRuston committed:
Update src/gnssro/gnss_tec_bufr2ioda.py

BenjaminRuston commented on src/gnssro/gnss_tec_bufr2ioda.py:
```diff
@@ -0,0 +1,375 @@
+    ('latitude', 'float'),
+    ('longitude', 'float'),
+    ('xECEFPosition', 'float'),
+    ('yECFPosition', 'float'),
+    ('zECFPosition', 'float'),
+    ('xECFPositionGNSS', 'float'),
+    ('yECFPositionGNSS', 'float'),
+    ('zECFPositionGNSS', 'float'),
```
```suggestion
    ('yECEFPosition', 'float'),
    ('zECEFPosition', 'float'),
    ('xECEFPositionGNSS', 'float'),
    ('yECEFPositionGNSS', 'float'),
    ('zECEFPositionGNSS', 'float'),
```

BenjaminRuston committed:
updating ECF to ECEF convention for Earth Center Earth Fixed

BenjaminRuston committed:
update convention to ECEF from ECF and enter stub for use of pyproj to compute sub-satellite point

BenjaminRuston committed:
add pyproj to transform from ECEF to lat lon

BenjaminRuston committed:
add height remove comments and unused variable

BenjaminRuston committed:
Merge branch 'develop' into feature/initial_tec_decoder

BenjaminRuston committed:
trigger CI

BenjaminRuston committed:
Merge branch 'feature/initial_tec_decoder' of https://github.com/jcsda-internal/ioda-converters into feature/initial_tec_decoder

BenjaminRuston committed:
trigger CI

huishao-r reviewed (commented):
It seems good to me. Based on feedback from SWPC, there might be some quality flag from the data file?  Should we check the raw data and see whether there is any QC related variable not yet included?

BenjaminRuston committed:
Merge branch 'develop' into feature/initial_tec_decoder
